JAX to HLO/jax_to_hlo.py

```diff
@@ -0,0 +1,199 @@
+(2019-06-13) best-effort.  It will mostly work, but it will occasionally break,
+and the XLA team won't (and in fact will be unable to) help.  One way to be sure
+it won't break is to use the same version of XLA to build the HLO as you use to
+run it.  The genrule above makes this easy.
+
+Implementation note: This script must be python2 compatible for now, because
+Google's genrules still run with python2, b/66712815.
+"""
+
+
+from ast import literal_eval
+import importlib
+import functools
+
+from absl import app
+from absl import flags
+import jax.api
+import jax.numpy as jnp
+from jax.lib import xla_client
+
+FLAGS = flags.FLAGS
+
+
+def jax_to_hlo(fn, input_shapes, constants=None):
+  """Converts a JAX function to an HLO module.
+
+  Args:
+    fn: Function to convert.
+    input_shapes: List of tuples (arg name, xla_client.Shape),
+      indicating the shapes of the arguments to fn.  The order of parameters in
+      the resulting XLA program will match the order in this list.
+    constants: Dict mapping function argument name to a Python value.  Specified
+      arguments these values as compile-time constants.
+
+  Returns:
+    A tuple (serialized_hlo_proto, hlo_text).
+  """
+  if not constants:
+    constants = {}
+
+  overlapping_args = {arg_name for arg_name, _ in input_shapes} & set(
+      constants.keys())
+  if overlapping_args:
+    raise ValueError(
+        'Arguments appear in both `input_shapes` and `constants`: %s' %
+        ', '.join(sorted(overlapping_args)))
+
+  args = []
+  for arg_name, shape in input_shapes:
+    if not shape.is_array():
+      raise ValueError('Shape %s is not an array, but currently only arrays '
+                       'are supported (i.e., no tuples).' % str(shape))
+
+    # Check that `shape` either doesn't have a layout or has the default layout.
+    #
+    # TODO(jlebar): This could be simpler if the Shape class exposed its layout,
+    # or if Shape exposed a function to unconditionally use the default layout.
+    shape_with_default_layout = xla_client.Shape.array_shape(
+        shape.xla_element_type(),
+        shape.dimensions()).with_major_to_minor_layout_if_absent()
+    if (shape.with_major_to_minor_layout_if_absent() !=
+        shape_with_default_layout):
+      raise ValueError('Shape %s has a non-default layout, but only '
+                       'the default layout is allowed.' % str(shape))
+
+    args.append(jnp.zeros(shape.dimensions(), dtype=shape.numpy_dtype()))
+
+  # Curry `constants` into the function.
+  fn_curried = functools.partial(fn, **constants)
+
+  # Wrapper that takes in args in the order of `input_shapes` and converts them
+  # to kwargs for calling `fn`.
+  def ordered_wrapper(*args):
+    arg_names = [arg_name for arg_name, _ in input_shapes]
+    return fn_curried(**dict(zip(arg_names, args)))
+
+  comp = jax.api.xla_computation(ordered_wrapper)(*args)
+  return comp # (comp.as_serialized_hlo_module_proto(), comp.as_hlo_text())
```

Comment:
I think this is the line you've changed (the pull request says line 138, which doesn't make sense).

Some other reader of this code (or yourself, in six weeks' time) would probably appreciate some commentary here about what this difference is for, rather than a commented-out expression with no context :)

Meanwhile I'm wondering if there's a straightforward way to mark this file as a copy of another file, so that the change can be automatically highlit.